### PR TITLE
Restructuring and cleaning for Era 2018

### PR DESCRIPTION
**Pull request by ram1123**
1. Change the directory hierarchy 
2. Removed all unwanted scripts. If we need some then we can always import later from our older versions.
3. Now, the script it working fine for m4l.
   1.  It will give error for other obs at last step. As it won't find the Datacard template to copy from. [runHZZFiducialXS.py#L227](https://github.com/ram1123/Fiducial_XS/blob/beb44cda9b4078285754e2bdf73a3c3083c919e9/runHZZFiducialXS.py#L227)

This PR should contain all the main changes from PR #1 from Tahir (@tjavaid )

### CODE REVIEW

File: RunEverything.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import os
+import argparse
+
+from python.collectInputs import collect
+from python.obsList import obsNamesBinsDict
+
+parser = argparse.ArgumentParser(description='Input arguments')
+parser.add_argument( '-s', dest='step', default=1, choices=[1, 2, 3, 4, 5], type=int, help='Which step to run')
+parser.add_argument( '-c', dest='channels', nargs="+",  default=["4mu", "4e", "2e2mu", "4l"], help='list of channels')
```
Comment (ram1123) [inline RunEverything.py:9]: For mass4l, 4channels for rest only 3 channels

File: RunEverything.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+for obsName, obsBin in obsNamesBinsDict.items():
+    print("="*51)
+    print("==> Obs: {:25}  {}".format(obsName,obsBin))
+    if (args.step == 1):
+        print("="*51)
+        print("Running the efficiencies step")
+        print("="*51)
+        for channel in args.channels:
+            print("==> channel: {}".format(channel))
+            command = 'nohup python -u efficiencyFactors.py -l -q -b --obsName="{obsName}" --obsBins="{obsBins}" -c "{channel}" >& log/effs_{obsName}_{channel}.log &'.format(
```
Comment (ram1123) [inline RunEverything.py:28]: Need to protect for submitting 100's of nohup submissions

File: efficiencyFactors.py [hunk 3]
```diff
@@ -208,104 +208,105 @@ def geteffs(channel, List, m4l_bins, m4l_low, m4l_high, obs_reco, obs_gen, obs_b
                 cutchan_gen_out  = "((GENZ_MomId[0]==25 && GENZ_MomId[1]==25 && GENZ_DaughtersId[0]==13 && GENZ_DaughtersId[1]==13) || (GENZ_MomId[0]==25 && GENZ_MomId[2]==25 && GENZ_DaughtersId[0]==13 && GENZ_DaughtersId[2]==13) || (GENZ_MomId[1]==25 && GENZ_MomId[2]==25 && GENZ_DaughtersId[1]==13 && GENZ_DaughtersId[2]==13))"
             if (channel == "2e2mu"):
                 cutchan_gen_out  = "((GENZ_MomId[0]==25 && (GENZ_DaughtersId[0]==11 || GENZ_DaughtersId[0]==13) && GENZ_MomId[1]==25 && (GENZ_DaughtersId[1]==11 || GENZ_DaughtersId[1]==13) && GENZ_DaughtersId[0]!=GENZ_DaughtersId[1]) || (GENZ_MomId[0]==25 && (GENZ_DaughtersId[0]==11 || GENZ_DaughtersId[0]==13) && GENZ_MomId[2]==25 && (GENZ_DaughtersId[2]==11 || GENZ_DaughtersId[2]==13) && GENZ_DaughtersId[0]!=GENZ_DaughtersId[2]) || (GENZ_MomId[1]==25 && (GENZ_DaughtersId[1]==11 || GENZ_DaughtersId[1]==13) && GENZ_MomId[2]==25 && (GENZ_DaughtersId[2]==11 || GENZ_DaughtersId[2]==13) && GENZ_DaughtersId[1]!=GENZ_DaughtersId[2]))"
- 
+
+        # FIXME: Why 10000 factor?
         if (recoweight=="totalWeight"): genweight = "10000.0*genWeight/"+str(sumw[Sample])
```
Comment (ram1123) [inline efficiencyFactors.py:213]: Why factor of 10000?

File: efficiencyFactors.py [hunk 4]
```diff
@@ -392,59 +393,59 @@ def geteffs(channel, List, m4l_bins, m4l_low, m4l_high, obs_reco, obs_gen, obs_b
...
-            
+
             if (channel=='2e2mu'):
-                CMS_zz4l_mean = RooFormulaVar("CMS_zz4l_mean_sig","CMS_zz4l_mean_sig","(124.771931539+(0.998407660986)*(@0-125))",RooArgList(MH))
-                CMS_zz4l_sigma = RooFormulaVar("CMS_zz4l_sigma_sig","CMS_zz4l_sigma_sig","(1.42180192725+(0.0100469936221)*(@0-125))",RooArgList(MH))                                                          
-                CMS_zz4l_alpha = RooFormulaVar("CMS_zz4l_alpha","CMS_zz4l_alpha","(0.927601791812)+(0*@0)",RooArgList(MH))
-                CMS_zz4l_n = RooFormulaVar("CMS_zz4l_n","CMS_zz4l_n","(3.18098806656+(-0.0188855891779)*(@0-125))",RooArgList(MH))
-                CMS_zz4l_alpha2 = RooFormulaVar("CMS_zz4l_alpha2","CMS_zz4l_alpha2","(1.56743054428)+(0*@0)",RooArgList(MH))
-                CMS_zz4l_n2 = RooFormulaVar("CMS_zz4l_n2","CMS_zz4l_n2","(3.98349053402+(0.0517139624607)*(@0-125))",RooArgList(MH))
+                CMS_zz4l_mean = RooFormulaVar("CMS_zz4l_mean_sig","CMS_zz4l_mean_sig","(124.260469656+(0.995095874123)*(@0-125))",RooArgList(MH))
```
Comment (ram1123) [inline efficiencyFactors.py:423]: copied these numbers from Tahir PR1

File: getUnc_Unc.py [hunk 5]
```diff
@@ -115,56 +115,53 @@ def getunc(channel, List, m4l_bins, m4l_low, m4l_high, obs_reco, obs_gen, obs_bi
...
         Histos[processBin+"fs"].Sumw2()
         if ("NNLOPS" in processBin):
             Tree[Sample].Draw("GENmass4l >> "+processBin+"fs","(nnloWeights[0])*("+cutchan_gen_out+")","goff")
         else:
             Tree[Sample].Draw("GENmass4l >> "+processBin+"fs","(qcdWeights[0])*("+cutchan_gen_out+")","goff")
-        
+
         if ("NNLOPS" in processBin):
-            #for i in range(9,36):
             for i in range(0,27):
```
Comment (ram1123) [inline getUnc_Unc.py:135]: Why number 27?

File: getUnc_Unc.py [hunk 6]
```diff
@@ -115,56 +115,53 @@ def getunc(channel, List, m4l_bins, m4l_low, m4l_high, obs_reco, obs_gen, obs_bi
...
-                #if (i==14 or i==16 or i==23 or i==25 or i==32 or i==34): continue
-                #if (i==5 or i==7 or i==14 or i==16 or i==23 or i==25): continue                                        
-                if (i==5 or i==7 or i==11 or i==14 or i==15 or i==16 or i==17 or i==19 or i==21 or i==22 or i==23 or i==25): continue 
+                if (i==5 or i==7 or i==11 or i==14 or i==15 or i==16 or i==17 or i==19 or i==21 or i==22 or i==23 or i==25): continue
```
Comment (ram1123) [inline getUnc_Unc.py:136]: How these numbers work?

File: getUnc_Unc.py [hunk 7]
```diff
@@ -181,44 +178,41 @@ def getunc(channel, List, m4l_bins, m4l_low, m4l_high, obs_reco, obs_gen, obs_bi
...
                 accerrstat = sqrt(acceptance[processBin]*(1-acceptance[processBin])/fsintegral)
-                qcderrup=1.0; qcderrdn=1.0; 
+                qcderrup=1.0; qcderrdn=1.0;
                 accerrup=1.0; accerrdn=1.0;
-                #for i in range(9,36):
-                for i in range(0,27):
-                    #if (i==14 or i==16 or i==23 or i==25 or i==32 or i==34): continue                                        
-                    #if (i==5 or i==7 or i==14 or i==16 or i==23 or i==25): continue                                        
-                    if (i==5 or i==7 or i==11 or i==14 or i==15 or i==16 or i==17 or i==19 or i==21 or i==22 or i==23 or i==25): continue 
+                for i in range(0,27): # FIXME: What is this number 27?
```
Comment (ram1123) [inline getUnc_Unc.py:190]: What is this number 27?

File: python/addConstrainedModel.py [hunk 8]
```diff
@@ -255,15 +254,17 @@ def parseOptions():
...
+
         print fState,obsName,'recobin',str(recobin),'jesSM',jesSM,'jesSMup',jesSMup,'jesSMdn',jesSMdn
 
-        
-        
+
+
+# FIXME: Why copying the .py file as _ORIG.py
+#        As I don't see any changes in the variables written to new .py file
 os.system('cp datacardInputs/inputs_sig_'+opt.OBSNAME+'.py datacardInputs/inputs_sig_'+opt.OBSNAME+'_ORIG.py')
```
Comment (ram1123) [inline python/addConstrainedModel.py:268]: We should remove this copy command.

File: runHZZFiducialXS.py [hunk 9]
```diff
@@ -297,19 +326,22 @@ def createAsimov(obsName, observableBins, modelName, resultsXS, physicalModel):
...
+        print("[INFO] command:\n\t{}".format(cmd))
         cmd = cmd +' -t -1 --saveWorkspace --saveToys'
+        # cmd = cmd + ' --out combineLogs' # FIXME: redirect log files from combine script to `combineLogs` directory
         #cmd += ' --X-rtd TMCSO_PseudoAsimov=1000000'
         #cmd += ' --freezeNuisanceGroups r4muBin0,r4eBin0,r2e2muBin0'
-        print cmd
+        print("[INFO] command:\n\t{}".format(cmd))
         output = processCmd(cmd)
-        processCmd('mv higgsCombine'+obsName+'.MultiDimFit.mH'+opt.ASIMOVMASS.rstrip('.0')+'.123456.root '+modelName+'_all_'+obsName+'_13TeV_Asimov_'+physicalModel+'.root',1)
+        if not os.path.isdir('combineLogs'): os.mkdir('combineLogs')
```
Comment (ram1123) [inline runHZZFiducialXS.py]: Update the directory name:

combineLogs -> combineOutputs

File: runHZZFiducialXS.py [hunk 10]
```diff
@@ -652,55 +684,64 @@ def runFiducialXS():
...
-        print output
-        asimovDataModelName = "SM_125"    
+        print (output)
+        asimovDataModelName = "SM_125"
         asimovPhysicalModel = "v2"
         produceDatacards(obsName, observableBins, asimovDataModelName, asimovPhysicalModel)
         resultsXS = createAsimov(obsName, observableBins, asimovDataModelName, resultsXS, asimovPhysicalModel)
         print "resultsXS: \n", resultsXS
-        
+    """
```
Comment (ram1123) [inline runHZZFiducialXS.py]: By mistake I commented bottom part. Will fix before merging this PR

**Comment by ram1123** [comment]
I checked the code with latest push in an independent directory and it seems that its working fine.

Now, everything is fine from my side. Let me know if you have any comments.

**Review by vukasinmilosevic — APPROVED** [review]
All looks good from the review perspective, will test it locally from my side before merging.

**Pull request merged**